pkg/controller/utils/ratelimitedeventhandler.go

```diff
@@ -50,3 +52,50 @@ var _ workqueue.RateLimitingInterface = &rateLimitedAddQueue{}
 func (q *rateLimitedAddQueue) Add(item interface{}) {
 	q.RateLimitingInterface.AddRateLimited(item)
 }
+
+// FIXME OCPBUGS-35282 - description
+func NewTypedRateLimitedUpdateEventHandlerClusterDeployment(typedEventHandler handler.TypedEventHandler[*hivev1.ClusterDeployment], shouldRateLimitFunc func(event.UpdateEvent) bool) handler.TypedEventHandler[*hivev1.ClusterDeployment] {
+	return &typedRateLimitedUpdateEventHandlerClusterDeployment{
+		TypedEventHandler: typedEventHandler,
+		shouldRateLimit:   shouldRateLimitFunc,
+	}
+}
+
+// FIXME how to not duplicate?
+func NewTypedRateLimitedUpdateEventHandlerDNSZone(typedEventHandler handler.TypedEventHandler[*hivev1.DNSZone], shouldRateLimitFunc func(event.UpdateEvent) bool) handler.TypedEventHandler[*hivev1.DNSZone] {
+	return &typedRateLimitedUpdateEventHandlerDNSZone{
+		TypedEventHandler: typedEventHandler,
+		shouldRateLimit:   shouldRateLimitFunc,
+	}
+}
+
+// FIXME how to not duplicate?
+func NewTypedRateLimitedUpdateEventHandlerMachinePool(typedEventHandler handler.TypedEventHandler[*hivev1.MachinePool], shouldRateLimitFunc func(event.UpdateEvent) bool) handler.TypedEventHandler[*hivev1.MachinePool] {
+	return &typedRateLimitedUpdateEventHandlerMachinePool{
+		TypedEventHandler: typedEventHandler,
+		shouldRateLimit:   shouldRateLimitFunc,
+	}
+}
+
+// rateLimitedUpdateEventHandler wraps the specified event handler such
+// that it will rate limit the incoming UPDATE events when the provided
+// shouldRateLimit function returns true.
+type typedRateLimitedUpdateEventHandlerClusterDeployment struct {
+	handler.TypedEventHandler[*hivev1.ClusterDeployment]
+
+	shouldRateLimit func(event.UpdateEvent) bool
+}
+
+type typedRateLimitedUpdateEventHandlerDNSZone struct {
+	handler.TypedEventHandler[*hivev1.DNSZone]
+
+	shouldRateLimit func(event.UpdateEvent) bool
+}
+
+type typedRateLimitedUpdateEventHandlerMachinePool struct {
+	handler.TypedEventHandler[*hivev1.MachinePool]
+
+	shouldRateLimit func(event.UpdateEvent) bool
+}
+
+// FXIME Add Create, Update, Delete methods for all?
```

Comment:
Yeah, no. Generics aren't supposed to make things worse. There's certainly a way around this. Unfortunately, I'm not very good with generics. @dlom has more experience, maybe he can have a look at these questions?

Reply:
Something like this should work:

```go
// NewRateLimitedUpdateEventHandler wraps the specified event handler inside a new
// event handler that will rate limit the incoming UPDATE events when the provided
// shouldRateLimit function returns true.
func NewRateLimitedUpdateEventHandler[T any](eventHandler handler.TypedEventHandler[T], shouldRateLimitFunc func(event.TypedUpdateEvent[T]) bool) handler.TypedEventHandler[T] {
	return &rateLimitedUpdateEventHandler[T]{
		TypedEventHandler: eventHandler,
		shouldRateLimit:   shouldRateLimitFunc,
	}
}

// rateLimitedUpdateEventHandler wraps the specified event handler such
// that it will rate limit the incoming UPDATE events when the provided
// shouldRateLimit function returns true.
type rateLimitedUpdateEventHandler[T any] struct {
	handler.TypedEventHandler[T]

	shouldRateLimit func(event.TypedUpdateEvent[T]) bool
}

var _ handler.TypedEventHandler[client.Object] = &rateLimitedUpdateEventHandler[client.Object]{}

// Update implements handler.EventHandler
func (h *rateLimitedUpdateEventHandler[T]) Update(ctx context.Context, e event.TypedUpdateEvent[T], q workqueue.RateLimitingInterface) {
	nq := q
	if h.shouldRateLimit(e) {
		nq = &rateLimitedAddQueue{q}
	}
	h.TypedEventHandler.Update(ctx, e, nq)
}
```